src/app/plugins/connectfour.plugin.ts

```diff
@@ -96,112 +100,211 @@ class ConnectFourGame {
   }
 
-  public getWinner() {
-    return this.winner;
-  }
-
-  public getTie() {
-    return this.tie;
+  public getGameOver() {
+    return this.gameOver;
   }
 
   public async move(col: number, msg: IMessage) {
     // Make move.
-    const successfulDrop: boolean = this.dropPiece(col);
-    if (!successfulDrop) {
+    if (!this._dropPiece(col)) {
       return;
     }
 
-    // Check for win.
-    if (this.checkWin()) {
-      this.winner = this.currentPlayer;
-    } else {
-      this.tie = this.checkTie();
+    await this._updateGameState(msg);
 
-      // Change players.
-      this.currentPlayer *= -1;
+    // Make Lion's move if the user is playing Lion.
+    if (!this.gameOver && this._playingLion) {
+      this._lionMove();
+      await this._updateGameState(msg);
     }
 
-    await msg.edit(this.showBoard());
-
     return;
   }
 
-  public dropPiece(col: number): boolean {
-    // Find first obstruction from top down in column.
-    let row = 0;
-    while (row < this._rows && this._board[row][col] === 0) {
-      row++;
+  // Turn on the beast.
+  private _lionMove() {
+    const moves: { col: number; val: number }[] = [];
+    for (let col = 0; col < this._cols; col++) {
+      if (!this._dropPiece(col)) {
+        continue;
+      }
+
+      moves.push({ col, val: this._evaluate(this.currentPlayer * -1, 0) });
+      this._removeTopPiece(col);
+    }
+
+    // Sort moves from best to worst.
+    moves.sort((a, b) => b.val - a.val);
+    const bestValue = moves[0].val;
+    const moveOptions = moves.filter(
+      (move) => Math.abs(bestValue - move.val) < this._getNormalizedDifficulty()
+    );
+    const randomMoveCol = moveOptions[Math.floor(Math.random() * moveOptions.length)].col;
+    this._dropPiece(randomMoveCol);
+  }
+
+  // We want the highest difficulty to allow for the least error.
+  private _getNormalizedDifficulty() {
+    return 1 / this._aiDifficulty;
+  }
+
+  // Evaluate the strength of the current board state,
+  // as it relates to Player 2's success.
+  // Return value is the average of all possible moves from this position
+  // Each move is evaluated on a score of [-4, 4]
+  private _evaluate(currentPlayer: number, depth: number) {
+    // If we have reached a win state, then the LAST move won.
+    if (this._checkWin()) {
+      return -4 * currentPlayer;
+    }
+    if (this._checkTie()) {
+      return 0;
+    }
+    // If we reached depth, then evaluate the board.
+    // We use a very simple evaluation: -(longest player one chain)
+    // This causes Lion to simply try to minimize Player 1's success,
+    // While our _checkWin case takes over when it's time to clinch victory.
+    if (depth === this._aiDepth) {
+      return this._longestChainOnBoard(1) - this._longestChainOnBoard(-1);
     }
-    // Subtract one to place piece one slot above obstruction.
+
+    const moves: number[] = [];
+    for (let col = 0; col < this._cols; col++) {
```

Comment:
Can we use a foreach?

Reply:
Why? We only need an index here, and the index needs to be column oriented. In order to pull columns indices that are semantically correct out of a forEach we would need

```javascript
this._board.forEach((_row, row) => row.forEach((_col, col) => {
    // do the thing.
})
```

This takes more time, and more memory.


If I just misinterpreted what you meant let me know.

Reply:
oh is it column oriented?
Then nevermind